### PR TITLE
Ability to clear cached potential energy computation in HMC

### PR DESCRIPTION
**Pull request by neerajprad**
Addresses #1926 - particularly, ability to clear the cache and recompute PE. 

I noticed that clearing the cache in warmup gives different results due to a different step size being learnt in adaptation (between steps 80 to 100), which is odd because this shouldn't affect the adaptation in any way. 

TODO:
 - [x] Debug why clearing the cache results in different adaptation results (step size changes after step 80 or so in warmup).

### CODE REVIEW

File: pyro/infer/mcmc/hmc.py [hunk 1]
```diff
@@ -270,13 +270,22 @@ def _cache(self, z, potential_energy, z_grads):
...
+
     def _fetch_from_cache(self):
         return self._z_last, self._potential_energy_last, self._z_grads_last
 
     def sample(self, params):
         z, potential_energy, z_grads = self._fetch_from_cache()
+        # recompute PE when cache is cleared
+        if z is None:
+            z = params
+            potential_energy = self.potential_fn(z)
```
Comment (fritzo) [inline pyro/infer/mcmc/hmc.py]: IIUC `z_grads` will still be `None` when passed into `velicity_verlet()` below. Is this intended?

Comment (fritzo) [reply pyro/infer/mcmc/hmc.py]: Ah I see this is ok :+1:

Comment (neerajprad) [reply pyro/infer/mcmc/hmc.py]: Yes, the integrator will compute it if it's None.

**Comment by fehiepsi** [comment]
I think that somehow MCMC just modifies `params` for the next step. Let me check where is the cause.

**Comment by neerajprad** [comment]
> the test does not use MCMC, so that might not be related. Let me check where is the cause.

@fehiepsi - the issue was with `find_reasonable_step_size` which uses the cache. I am currently recomputing here as well, though it might be simpler to just not use cache in this function since it is probably called only a select number of times.

**Comment by fehiepsi** [comment]
Yeah, that's right!

**Pull request merged**